### ISSUE TITLE
GravityView Shortcode causes the query to load twice

### ISSUE THREAD

**Issue by shehrozsheikh** [GravityKit/GravityView#2089]
**Describe the Issue**

Hi @zackkatz, I work at [Zao.is](https://zao.is), and one of our clients [EditorNinja](https://editorninja.com) requested us to look at their slow page loading issue. He was using the [GravityView](https://docs.gravitykit.com/article/73-using-the-shortcode) shortcode on that page.

I debugged it using Query Monitor and found some issues with this query, it was running twice:

`SELECT SQL_CALC_FOUND_ROWS DISTINCT `t1`.`id`
FROM `wp_gf_entry` AS `t1`
LEFT JOIN `wp_gf_entry_meta` AS `m2`
ON (`m2`.`entry_id` = `t1`.`id`
AND `m2`.`meta_key` = '34')
LEFT JOIN `wp_gf_entry_meta` AS `m3`
ON (`m3`.`entry_id` = `t1`.`id`
AND `m3`.`meta_key` = '21')
LEFT JOIN `wp_gf_entry_meta` AS `m4`
ON (`m4`.`entry_id` = `t1`.`id`
AND `m4`.`meta_key` = '60')
LEFT JOIN `wp_gf_entry_meta` AS `m5`
ON (`m5`.`entry_id` = `t1`.`id`
AND `m5`.`meta_key` = '7')
LEFT JOIN `wp_gf_entry_meta` AS `m6`
ON (`m6`.`entry_id` = `t1`.`id`
AND `m6`.`meta_key` = '16')
LEFT JOIN `wp_gf_entry_meta` AS `m7`
ON (`m7`.`entry_id` = `t1`.`id`
AND `m7`.`meta_key` = '15')
LEFT JOIN `wp_gf_entry_meta` AS `o8`
ON (`o8`.`entry_id` = `t1`.`id`
AND `o8`.`meta_key` = '8')
LEFT JOIN `wp_gf_entry_meta` AS `o9`
ON (`o9`.`entry_id` = `t1`.`id`
AND `o9`.`meta_key` = '18')
WHERE (`t1`.`form_id` IN (9)
AND (`t1`.`status` = 'active'
AND ((`t1`.`created_by` = '1'
OR (`m2`.`meta_key` = '34'
AND `m2`.`meta_value` = 'john@editorninja.com')
OR (`m3`.`meta_key` = '21'
AND `m3`.`meta_value` = 'john@editorninja.com')
OR `m4`.`meta_key` = '60'
AND CAST(`m4`.`meta_value` AS DECIMAL(65, 6)) = '1'
OR (`m5`.`meta_key` = '7'
AND `m5`.`meta_value` = 'john@editorninja.com'))
AND ((NOT EXISTS(SELECT 1
FROM `wp_gf_entry_meta`
WHERE `meta_key` = '16'
AND `entry_id` = `t1`.`id`)
OR (`m6`.`meta_key` = '16'
AND `m6`.`meta_value` != 'Completed'))
OR NOT EXISTS(SELECT 1
FROM `wp_gf_entry_meta`
WHERE (`meta_key` LIKE '16.%'
OR `meta_key` = 16)
AND `entry_id` = `t1`.`id`))
AND (((NOT EXISTS(SELECT 1
FROM `wp_gf_entry_meta`
WHERE `meta_key` = '15'
AND `entry_id` = `t1`.`id`)
OR (`m7`.`meta_key` = '15'
AND `m7`.`meta_value` != 'Returned'))
OR NOT EXISTS(SELECT 1
FROM `wp_gf_entry_meta`
WHERE (`meta_key` LIKE '15.%'
OR `meta_key` = 15)
AND `entry_id` = `t1`.`id`))
OR ((NOT EXISTS(SELECT 1
FROM `wp_gf_entry_meta`
WHERE `meta_key` = '15'
AND `entry_id` = `t1`.`id`)
OR (`m7`.`meta_key` = '15'
AND `m7`.`meta_value` != 'Approved'))
OR NOT EXISTS(SELECT 1
FROM `wp_gf_entry_meta`
WHERE (`meta_key` LIKE '15.%'
OR `meta_key` = 15)
AND `entry_id` = `t1`.`id`))
OR ((NOT EXISTS(SELECT 1
FROM `wp_gf_entry_meta`
WHERE `meta_key` = '15'
AND `entry_id` = `t1`.`id`)
OR (`m7`.`meta_key` = '15'
AND `m7`.`meta_value` != 'Duplicate'))
OR NOT EXISTS(SELECT 1
FROM `wp_gf_entry_meta`
WHERE (`meta_key` LIKE '15.%'
OR `meta_key` = 15)
AND `entry_id` = `t1`.`id`))))))
ORDER BY `o8`.`meta_value` DESC, `o9`.`meta_value` DESC
LIMIT 25`

It works fine and runs only once when we preview the single view page from Edit View > Preview, the issue is only with the shortcode.

<img width="1511" alt="image" src="https://github.com/user-attachments/assets/85b0c373-2540-46ec-ba4d-41b0095c282f">

I verified with Diffchecker that the same query is being executed twice. When I commented out lines 139 and 140 in class-gv-shortcode-gravityview.php, the issue was resolved. I'm not entirely sure of the purpose of these lines, so there might be a better solution that your team can investigate but I wanted to bring this to your attention. I don't believe the issue is related to the View settings, as it works correctly when previewing a single View page without using the shortcode.

Here is the detailed [LOOM VIDEO](https://www.loom.com/share/d9da3efb9ae440f780ec2a28f2b49928?sid=afcef242-8644-46dd-b35d-86263287ef81) to describe it in a better way.

Initially, the page loading time was 16-17 seconds, I brought it down to 5-6 seconds by adding this indexing on the database table, so maybe you can create a small tool for it to create the indexing -> 

```
CREATE INDEX idx_entry_meta_value ON wp_gf_entry_meta (entry_id, meta_key, meta_value(191));
CREATE INDEX idx_meta_entry ON wp_gf_entry_meta (meta_key, entry_id);
CREATE INDEX idx_form_status_created_by ON wp_gf_entry (form_id, status, created_by, id);
```
That 5-6 seconds further got reduced to 2-3 seconds when I commented out these lines. That client has 5000+ records on that page, but it has a setting of showing only 25 per page.

--------

### Additional information

**This issue happens in the following settings:**

**View type:** List

_Please try to re-create the issue in each rendering mode. This helps debug the problem._

- [ ] View
- [x] Embedded via shortcode
- [ ] Embedded via oEmbed

**Comment by mrcasual** [comment]
@shehrozsheikh, thank you for bringing this to our attention. We'll review the findings and see how best to resolve the issue.

**Comment by shehrozsheikh** [comment]
> @shehrozsheikh, thank you for bringing this to our attention. We'll review the findings and see how best to resolve the issue given the various constraints (e.g., we rely on the `GF_Query` class, not all hosts allow creating indices, etc.)

Sure @mrcasual, thanks for the prompt response. They are using Kinsta, and I created indexing manually by going into MySQL so I think you can still make a tool and disable/enable it based on the permissions from the hosting.

Also, let me know if you need the staging credentials for further debugging or help. You can email me at shehroz@zao.is.

**Comment by mrcasual** [comment]
@shehrozsheikh, just to confirm, you disabled the following 2 lines?

https://github.com/GravityKit/GravityView/blob/d33ca4d0490f0891de4be5288d0ed5e1ea86438a/future/includes/class-gv-shortcode-gravityview.php#L139-L140

**Comment by sahmed2-godaddy** [comment]
Yes exactly, @mrcasual.

**Comment by rafaehlers** [comment]
@mrcasual Has this been done?

**Comment by mrcasual** [comment]
> @mrcasual Has this been done?

@rafaehlers, released in 2.26: 

`Caching of View entries to prevent unnecessary database queries.`
